lambdas/es/indexer/index.py

```diff
@@ -1,74 +1,220 @@
+                # These environment variables are automatically set by Lambda
+                aws_access_key=os.environ["AWS_ACCESS_KEY_ID"],
+                aws_secret_access_key=os.environ["AWS_SECRET_ACCESS_KEY"],
+                aws_token=os.environ["AWS_SESSION_TOKEN"],
+                aws_host=elastic_host,
+                aws_region=boto3.session.Session().region_name,
+                aws_service="es"
+            )
+
+            time_remaining = get_time_remaining(self.context)
+
+            elastic = Elasticsearch(
+                hosts=[{"host": elastic_host, "port": 443}],
+                http_auth=awsauth,
+                max_backoff=time_remaining,
+                # Give ES time to repsond when under laod
+                timeout=ELASTIC_TIMEOUT,
+                use_ssl=True,
+                verify_certs=True,
+                connection_class=RequestsHttpConnection
+            )
+
+            _, errors = bulk(
+                elastic,
+                iter(self.queue),
+                # Some magic numbers to reduce memory pressure
+                # e.g. see https://github.com/wagtail/wagtail/issues/4554
+                # The stated default is max_chunk_bytes=10485760, but with default
+                # ES will still return an exception stating that the very
+                # same request size limit has been exceeded
+                chunk_size=100,
+                max_chunk_bytes=CHUNK_LIMIT_BYTES,
+                # number of retries for 429 (too many requests only)
+                # all other errors handled by our code
+                max_retries=RETRY_429,
+                # we'll process errors on our own
+                raise_on_error=False,
+                raise_on_exception=False
+            )
+            # reset the size count
+            self.size = 0
+            # Retry only if this is a first-generation queue
+            # (prevents infinite regress on failing documents)
+            if self.retry_errors:
+                # this is a second genration queue, so don't let it retry
+                # anything other than 429s
+                error_queue = DocumentQueue(self.context, retry_errors=False)
+                for error in errors:
+                    print(error)
+                    # can be dict or string *sigh*
+                    inner = error.get("index", {})
+                    error_info = inner.get("error")
+                    doc_id = inner.get("_id")
+
+                    if isinstance(error_info, dict):
+                        error_type = error_info.get("type", "")
+                        if 'mapper_parsing_exception' in error_type:
+                            replay = next(doc for doc in self.queue if doc["_id"] == doc_id)
+                            replay['user_meta'] = replay['system'] = {}
+                            error_queue.append_document(replay)
+                # recursive but never goes more than one level deep
```

Comment:
Still, the recursion bothers me a bit. What about moving everything but the error logic into a helper function, and then:
```
_, errors = self._send_all_helper(self.queue)
if errors:
  error_queue = ...
  self._send_all_helper(error_queue)
```
?